[DDW-409] New mnemonic input

przemyslaw-wlodek opened the pull request:
This PR implements a new design of a Mnemonic Input component.

## Screenshots

Before:

<img width="648" alt="image" src="https://user-images.githubusercontent.com/17825044/171393091-3b19d751-d841-4d97-8604-f6740206f06c.png">

<img width="646" alt="image" src="https://user-images.githubusercontent.com/17825044/171448895-a5bfc7a4-f860-4ad9-8716-0c6587ab6baa.png">

After:

<img width="648" alt="image" src="https://user-images.githubusercontent.com/17825044/171392982-2fb1afa1-05c5-4ea3-9219-255c6c0c2204.png">

<img width="645" alt="image" src="https://user-images.githubusercontent.com/17825044/171449219-a422abd2-d798-4257-806a-ee2743a9e31c.png">

## Testing Checklist

- [Slack QA thread](https://input-output-rnd.slack.com/archives/GGKFXSKC6/p1654099023110429)
- [ ] Test

---
## Test Scenarios



**Scenario 1 - Validate Creation Wizard - Step 3**
```gherkin
Given that I am on the step 3 of the wallet creation wizard
And I see the recovery phrase for the new wallet
Then I can see each word is numbered
And they are shown in 3 columns
```

**Scenario 2 - Validate |Recovery phrase dialog|**
```gherkin
Given that I am on the |Recovery phrase dialog|
When I validate the screen
Then I can see there is a dedicated numbered input for each word
And these are shown in 3 columns
And the amount of inputs matches the number of words for the type of wallet
```

**Scenario 3 - Paste a valid recovery phrase**
```gherkin
Given that I am on the |Recovery phrase dialog|
And I have copied a valid recovery phrase with the correct amount of words
When I position the cursor on any input
And I paste the recovery phrase
Then I can see that the recovery phrase is pasted 
And these are inserted on the corresponding order on each input
And over the recovery phrase I can see a message that informs all the words have been inserted
```

**Scenario 4 - Paste a single word**
```gherkin
Given that I am on the |Recovery phrase dialog|
And I have copied a single word from the recovery phrase
When I position the cursor on any input
And I paste the word
Then I can see the word is inserted correctly on the input
And the cursor moves to the next input
```

**Scenario 5 - Paste a wrong recovery phrase**
```gherkin
Given that I am on the |Recovery phrase dialog|
And I have copied an invalid recovery phrase with the correct amount of words
When I position the cursor on any input
And I paste the recovery phrase
Then I can see that the recovery phrase is pasted 
And these are inserted on the corresponding order on each input
And over the recovery phrase I can see a message that informs that the recovery phrase is invalid
```

**Scenario 6 - Paste a short or long recovery phrase**
```gherkin
Given that I am on the |Recovery phrase dialog|
And I have copied an invalid recovery phrase with the incorrect amount of words
When I position the cursor on any input
And I paste the recovery phrase
Then I can see that the recovery phrase is not pasted
```

**Scenarion 7 - Insert a valid recovery phrase manually**
```gherkin
Given that I am on the |Recovery phrase dialog|
And I have a valid recovery phrase with the correct amount of words
When I type each word on the inputs
Then I can see these are inserted correctly 
And each time I insert one valid word, the cursor moves to the next input
```

**Scenario 8 - Insert an invalid recovery phrase manually**
```gherkin
Given that I am on the |Recovery phrase dialog|
And I have a valid recovery phrase with the correct amount of words
When I type each word on the inputs
Then I can see these are inserted correctly 
And after all were entered, over the recovery phrase I can see a message that informs that the recovery phrase is invalid
```

**Scenario 10 - Delete a single word from the recovery phrase**
```gherkin
Given that I am on the |Recovery phrase dialog|
And I have entered a valid recovery phrase
When I delete a single word
Then I can see the counter over the recovery phrase indicates the actual value
And the submit button is disabled
```

**Scenario 11 - Insert an invalid word**
```gherkin
Given that I am on the |Recovery phrase dialog|
And I entered an invalid word 
And I switch to the next input
Then I can see the input with the invalid word is outlined in red
```

**Scenario 12 - Validate dropdown**
```gherkin
Given that I am on the |Recovery phrase dialog|
When I click on an input
Then I can see a dropdown with a list of valid words sorted alphabetically is displayed
And if I type a character the list is updated showing matching results
```

**Scenario 13 - Pasting mnemonic words**
```gherkin
Given that I am on the |Recovery phrase dialog|
And I have a valid recovery phrase with the correct amount of words
And the words are separated by only a space
And I have copied the recovery phrase into the clipboard
And I have clicked on an input field
When I paste the words
Then all of the pasted words appear according to the order they are written
And no input field is left empty
```

**Scenario 14 - Single input error (underline)**
```gherkin
Given that I am on the |Recovery phrase dialog|
And I have provided an incorrect word to one of the inputs
Then a red underline should be displayed under that input
```

**Scenario 15 - Submission with wrong phrase**
```gherkin
Given that I am on the |Recovery phrase dialog|
And I have provided an incorrect recovery phrase
And I submitted that phrase
Then error is displayed in top-right corner of the Mnemonic Input component
And I change one of the inputs to a different word
Then submission should be allowed again
```

```gherkin
|Recovery phrase dialog|
|Wallet Creation - Step 4|
|Wallet Restoration - Step 3|
|Wallet Recovery Phrase Verification - Step 2|
```

---

## Review Checklist

### Basics

- [x] PR assigned to the PR author(s)
- [x] `input-output-hk/daedalus-dev` and `input-output-hk/daedalus-qa` assigned as PR reviewers
- [ ] If there are UI changes, Alexander Rukin assigned as an additional reviewer
- [ ] All visual regression testing has been reviewed (assign `run Chromatic` label to PR to trigger the run)
- [x] PR has appropriate labels (`release-vNext`, `feature`/`bug`/`chore`, `WIP`)
- [x] PR link is added to a Jira ticket, ticket moved to In Review
- [x] PR is updated to the most recent version of the target branch (and there are no conflicts)
- [x] PR has a good description that summarizes all changes
- [x] PR contains screenshots (in case of UI changes)
- [x] CHANGELOG entry has been added to the top of the appropriate section (_Features_, _Fixes_, _Chores_) and is linked to the correct PR on GitHub
- [x] There are no missing translations (running `yarn manage:translations` produces no changes)
- [x] Text changes are proofread and approved (Jane Wild / Amy Reeve)
- [x] Japanese text changes are proofread and approved (Junko Oda)
- [x] Storybook works and no stories are broken (`yarn storybook`)
- [ ] In case of dependency changes `yarn.lock` file is updated

### Code Quality

- [ ] Important parts of the code are properly commented and documented
- [ ] Code is properly typed with typescript types
- [ ] React components are split-up enough to avoid unnecessary re-renderings
- [ ] Any code that only works in main process is neatly separated from components

### Testing

- [ ] New feature/change is covered by acceptance tests
- [ ] New feature/change is manually tested and approved by QA team
- [ ] All existing acceptance tests are still up-to-date
- [ ] New feature/change is covered by Daedalus Testing scenario
- [ ] All existing Daedalus Testing scenarios are still up-to-date

### After Review

- [ ] Update Slack QA thread by marking it with a green checkmark

lucas-barros commented on source/renderer/app/components/wallet/mnemonic-input/MnemonicAutocompleteContainer.tsx:
```diff
@@ -0,0 +1,169 @@
+  noResultsMessage,
+  inputRef,
+}: MnemonicAutocompleteContainerProps) => {
+  const initialState = useMemo(
```
I think useMemo is not necessary as you can have the same result using [lazy-initial-state](https://reactjs.org/docs/hooks-reference.html#lazy-initial-state)

lucas-barros replied:
Never mind, I see it is used again on reset.

renanvalentin commented:
Great job @przemyslaw-wlodek ! I found a small issue when testing copy/paste behavior:

1- Select recovery of 24 words
2- Focus on field#1
3- Paste recovery phrase
4- Go to field#12
5- Paste recovery phrase

**Expected behavior**
I would expect to have recovery phrase  applied again

**Actual behavior**
field#12 contains all recovery phrase

renanvalentin commented on source/renderer/app/components/wallet/mnemonic-input/MnemonicAutocompleteLayout.tsx:
```diff
@@ -0,0 +1,114 @@
+  onSelect: (value: string) => void;
+  onBlur: () => void;
+  onPaste: ClipboardEventHandler<HTMLInputElement>;
+  toggleMouseLocation: () => void;
```
should this be prefixed with `on` as well?

przemyslaw-wlodek replied:
I wanted consistency with react-polymorph `Options` props. Of course, we can change that but still, it needs to be updated in RP :)

renanvalentin commented on source/renderer/app/components/wallet/mnemonic-input/MnemonicAutocompleteContainer.tsx:
```diff
@@ -0,0 +1,169 @@
+
+  useEffect(() => {
+    if (value) {
+      setState((prevState) => ({ ...prevState, inputValue: value }));
```
Can we make this component stateless by removing this state and let consumers to hold handle its value?

miorsufianiohk commented:
Hi @przemyslaw-wlodek ,

This is my findings on for built 21866. I am aware that you are updating the document to make it clearer for everybody but I am raising these suggestions now in case I forget:

1) Once I entered a few words (or all the mnemonic words), in order to have everything blank, I had to delete one word at a time. Is it possible to delete all the words at once? 
2) I am unsure if this is by design. When I entered the words, clicking on them would always bring the suggestions "abandon ability able ...." even when the words were not listed in the suggested list. For example when I had the word "rare", clicking on "rare" would bring up the "abandon ability able.." list instead of "rabbit raccoon race...." 

Please see [video recording](https://drive.google.com/file/d/1qNSpTM0f_F4Mi1fs5gbofhzNP0sFDC8I/view?usp=sharing) for further info

DominikGuzei commented on source/renderer/app/components/wallet/mnemonic-input/MnemonicAutocompleteLayout.tsx:
```diff
@@ -0,0 +1,114 @@
+
+  value: string;
+  ordinalNumber: number;
+  disabled: boolean;
```
We should make sure that this component will not trigger any callbacks when it is disabled (to remove the `useCallback` + checks in the container)

DominikGuzei commented on source/renderer/app/components/wallet/mnemonic-input/MnemonicAutocompleteContainer.tsx:
```diff
@@ -0,0 +1,169 @@
+  const optionsRef = useRef();
+  const suggestionsRef = useRef();
+
+  const toggleOpen = useCallback(() => {
```
Why are we using `useCallback` here instead of toggling the state in an event handler directly? We don't need to memorize anything (same for all other cases in this function)

DominikGuzei reviewed (changes requested):
Good work 👍 Here is also some UX feedback:

1. Input fields error state when clicking and not entering anything:
![grafik](https://user-images.githubusercontent.com/172414/173344118-85276501-4d8e-40ea-88d7-a9c931bdd8ea.png)
As we already have the word count and everything else greyed out it might not be necessary to turn the borders red?

2. The cursor can be moved forward by hitting TAB but it's not possible to go back to the previous field with the keyboard (might not be necessary though)

przemyslaw-wlodek commented:
> Hi @przemyslaw-wlodek ,
> 
> These are my findings for built 21866. I am aware that you are updating the document to make it clearer for everybody but I am raising these suggestions now in case I forget:
> 
> 1. Once I entered a few words (or all the mnemonic words), in order to have everything blank, I had to delete one word at a time. Is it possible to delete all the words at once?
> 2. I am unsure if this is by design. When I entered the words, clicking on them would always bring the suggestions "abandon ability able ...." even when the words were not listed in the suggested list. For example when I had the word "rare", clicking on "rare" would bring up the "abandon ability able.." list instead of "rabbit raccoon race...."
> 3. I am not sure what the 23 of 24 words entered signify? As from [this screenshot](https://drive.google.com/file/d/1KnBRWLBnqh8PrvCD-ywJ4remj-TLAnjf/view?usp=sharing) there are blanks in position 12 and 23.
> 
> Please see [video recording](https://drive.google.com/file/d/1qNSpTM0f_F4Mi1fs5gbofhzNP0sFDC8I/view?usp=sharing) for further info

@miorsufianiohk I believe that all of the issues listed above are fixed for now. Thank you!
CC: @renanvalentin - this also addresses your concerns

przemyslaw-wlodek commented:
> Good work 👍 Here is also some UX feedback:
> 
> 1. Input fields error state when clicking and not entering anything:
>    ![grafik](https://user-images.githubusercontent.com/172414/173344118-85276501-4d8e-40ea-88d7-a9c931bdd8ea.png)
>    As we already have the word count and everything else greyed out it might not be necessary to turn the borders red?
> 2. The cursor can be moved forward by hitting TAB but it's not possible to go back to the previous field with the keyboard (might not be necessary though)

Thank you @DominikGuzei! UX requirements were slightly different (as discussed offline) so the error states can remain as-is for now. Regarding navigation, keyboard arrows are used to navigate through options popover so I'd stick to `Tab` or `Shift+Tab` to navigate through the fields. 🙂

przemyslaw-wlodek commented on source/renderer/app/components/wallet/mnemonic-input/MnemonicAutocompleteContainer.tsx:
```diff
@@ -0,0 +1,169 @@
+      mouseIsOverOptions: false,
+      blurred: false,
+    }),
+    []
```
Thanks for pointing it out! 🙂 

1. We don't want to update it when `value` changes as this is the `initialState` used for `reset` functionality.
2. Technically `options` should not change in time by design, but to be politically correct I can update the memo state if they change. - What do you think @renanvalentin ? 😄

renanvalentin replied:
I guess we can keep it as it is

miorsufianiohk commented:
Hi @przemyslaw-wlodek ,

This is an interim report for build 21907 as I am still checking. Reporting this early so you could take action together with the previously reported issues by other QAs. I will add into this list if I find further issues and let you know.

- When there are line breaks between the pasted words, the words will appear continuously  on the chosen position. See [this screenshot](https://drive.google.com/file/d/1b8AXyuUgpJfG4S0HlUu1J4CF35LKy_nq/view?usp=sharing) 
Example mnemonic words with line break like this:
dawn
cake
inquiry
...

However, when there’s no line break, the words appear as expected. 
Example mnemonic words without line break like this:
dawn cake inquiry ...

- Pressing backspace on a word will delete the word but the cursor doesn’t move to previous position making it hard to delete multiple words just by pressing the backspace. Please note that in current Daedalus production, user could delete multiple words easily

- @gabriela-ponce has raised this issue before: I wanted to type “winner” but upon typing the word “win”, the word “win” was accepted and the cursor went to the next line.

- Not able to delete multiple words in one go. Would be nice to be able to delete multiple words at the same time or at least select all of the words with Cmd+A or Ctr+A. In current Daedalus production, you could delete multiple words just by hitting the backspace button once.

przemyslaw-wlodek commented on source/renderer/app/components/wallet/mnemonic-input/MnemonicAutocompleteContainer.tsx:
```diff
@@ -0,0 +1,169 @@
+      let selectedOption = '';
+      let isOpen = true;
+
+      if (options.includes(inputValue)) {
```
As discussed offline - we probably don't want that for mnemonics. This also conflicts a bit with functionality that automatically switches to the next input after you provide a correct word (which might be the part of another search in this case)

przemyslaw-wlodek commented:
Hey @miorsufianiohk :)
I've spoken with @danielmain and we've come up with this proposal:

> * When there are line breaks between the pasted words, the words will appear continuously  on the chosen position. See [this screenshot](https://drive.google.com/file/d/1b8AXyuUgpJfG4S0HlUu1J4CF35LKy_nq/view?usp=sharing)

We do not support line breaks, only a single space ` ` is allowed. Sorry that it wasn't specified in the AC.

> * Pressing backspace on a word will delete the word but the cursor doesn’t move to previous position making it hard to delete multiple words just by pressing the backspace. Please note that in current Daedalus production, user could delete multiple words easily
> * Not able to delete multiple words in one go. Would be nice to be able to delete multiple words at the same time or at least select all of the words with Cmd+A or Ctr+A. In current Daedalus production, you could delete multiple words just by hitting the backspace button once.

In some forms, there's a "Clear" button. This could be addressed on that level if needed, but I am not sure if that's a common case from a real user perspective. I do not recommend switching inputs while deleting a word (but we can discuss it with @danielmain / @daniloprates) :)

> * @gabriela-ponce has raised this issue before: I wanted to type “winner” but upon typing the word “win”, the word “win” was accepted and the cursor went to the next line.

This one is already fixed 😌

miorsufianiohk commented:
Hi @przemyslaw-wlodek ,

Thanks for fixing the above issues. Listed are some of the issues that I found on build 21960 (I will add more here and let you know if there's any new issue)

I was able to click on “Check Recovery Phrase” even when not all of the words were populated Seems a bit odd :thinking_face: . [See recording here](https://drive.google.com/file/d/1v0_0YhPVUbUqWhFEKek7QcfJwz6aRBPy/view?usp=sharing)

miorsufianiohk commented:
Hi @przemyslaw-wlodek,
 
This is the report for build 21968

- When I tried to create a new wallet, upon pasting the recovery phrase, the “Confirm” button was still disabled. [See video](https://drive.google.com/file/d/1v6vqELfROgzQ0oC7w3um2BvfLAKjuRgZ/view?usp=sharing).
- I am still able to enter symbols and numbers. [See screenshot](https://drive.google.com/file/d/11-NMAVu5uHBcQsg6P4vPVyBr0jFlNmbN/view?usp=sharing)

miorsufianiohk commented:
Hi @przemyslaw-wlodek ,

This is the findings for build 21997.

- I was able to reproduce all of the issues found on build 21968 as [reported here](https://github.com/input-output-hk/daedalus/pull/2979#issuecomment-1172168254)

- I also found an issue during recovery phrase verification. When I typed an invalid phrase, upon clicking "Verify" button, spinner continued spinning and console displayed error unable to "...retrieve the walletid". It should have displayed an error to the user indicating the phrase was invalid (ps: For further info,  Create new wallet and Restore wallet worked as expected). Please [see video](https://drive.google.com/file/d/10ArvE_UrAGQFfgjoxmuYDVRqsiCyIQFb/view?usp=sharing) for more info.

przemyslaw-wlodek commented:
> Hi @przemyslaw-wlodek ,
> 
> This is the findings for build 21997.
> 
> * I was able to reproduce all of the issues found on build 21968 as [reported here](https://github.com/input-output-hk/daedalus/pull/2979#issuecomment-1172168254)
> * I also found an issue during recovery phrase verification. When I typed an invalid phrase, upon clicking "Verify" button, spinner continued spinning and console displayed error unable to "...retrieve the walletid". It should have displayed an error to the user indicating the phrase was invalid (ps: For further info,  Create new wallet and Restore wallet worked as expected). Please [see video](https://drive.google.com/file/d/10ArvE_UrAGQFfgjoxmuYDVRqsiCyIQFb/view?usp=sharing) for more info.

Thank you @miorsufianiohk !

1. I am not able to reproduce them. Let's have a call.
2. Good catch! It's fixed now. The issue wasn't produced by my changes but it was revealed by them. 😅

miorsufianiohk commented:
Hi @przemyslaw-wlodek ,

These are the findings for build 22021

Overall, great job so far 👍 .

As discussed in our call:

- When I tried to create a new wallet, upon pasting a valid recovery phrase which was not from the newly created wallet, the “Confirm” button was still disabled. [See video](https://drive.google.com/file/d/1v6vqELfROgzQ0oC7w3um2BvfLAKjuRgZ/view?usp=sharing). (not yet fixed)
- ~I am still able to enter symbols and numbers. [See screenshot](https://drive.google.com/file/d/11-NMAVu5uHBcQsg6P4vPVyBr0jFlNmbN/view?usp=sharing)~ (this is expected behaviour)
- ~I also found an issue during recovery phrase verification. When I typed an invalid phrase, upon clicking "Verify" button, spinner continued spinning and console displayed error unable to "...retrieve the walletid". It should have displayed an error to the user indicating the phrase was invalid (ps: For further info, Create new wallet and Restore wallet worked as expected). Please [see video](https://drive.google.com/file/d/10ArvE_UrAGQFfgjoxmuYDVRqsiCyIQFb/view?usp=sharing) for more info.~ (fixed)

miorsufianiohk commented:
Hi @przemyslaw-wlodek ,

These are the findings for build 22026.

Great jobs on fixing the issue in the [previous report](https://github.com/input-output-hk/daedalus/pull/2979#issuecomment-1178007758)

However, as discussed in our chat, I found an issue that might have regressed in the wallet recovery verification

- For Yoroi Byron and Yoroi Shelley. Number of words were 15 but on recovery phrase dialog displayed 24

I also found an issue that potentially exists in production which I am verifying at the moment. If this is the case, I will create a separate card once verified
- Byron Paper Wallet. Number of words were 24 but on recovery phrase dialog displayed 12

Plese watch [video here](https://drive.google.com/file/d/14sUl9quoQI-lUOggH3n71OAn2lWkRZhs/view?usp=sharing) for further info

miorsufianiohk reviewed (approved):
LGTM. Great work @przemyslaw-wlodek 🎉 . Tested on 22080

renanvalentin reviewed (approved):
Great job @przemyslaw-wlodek !

lucas-barros commented on source/renderer/app/i18n/locales/ja-JP.json:
```diff
@@ -1114,7 +1108,7 @@
   "wallet.settings.passwordNotSet": "パスワードが設定されていません",
   "wallet.settings.recoveryPhraseInputNoResults": "該当するフレーズは見つかりませんでした",
-  "wallet.settings.recoveryPhraseInputPlaceholder": "{wordNumber}番目の単語を入力",
+  "wallet.settings.recoveryPhraseInputPlaceholder": "!!!Enter word #{wordNumber}",
```
@przemyslaw-wlodek Is this expected?

przemyslaw-wlodek replied:
Totally not. Thank you @lucas-barros!
